src/test/java/com/gluonhq/richtext/undo/CmdManagerTests.java

```diff
@@ -52,6 +54,109 @@ public void redoRules() {
+        Assertions.assertTrue(commander.isRedoStackEmpty());
+    }
+
+    @Test
+    @DisplayName("undo stack must be empty after a undo operation")
+    public void undoStackEmptyAfterUndoExecution() {
+        StringBuilder text = new StringBuilder("Text");
+        CommandManager<StringBuilder> commander = new CommandManager<>(text);
+        commander.execute(new TestCommand());
+        commander.undo();
+        Assertions.assertTrue(commander.isUndoStackEmpty());
+    }
+
+    @Test
+    @DisplayName("redo stack must not be empty after undo operation")
+    public void redoStackNotEmptyAfterUndoOperation() {
+        StringBuilder text = new StringBuilder("Text");
+        CommandManager<StringBuilder> commander = new CommandManager<>(text);
+        commander.execute(new TestCommand());
+        commander.undo();
+        Assertions.assertFalse(commander.isRedoStackEmpty());
+    }
+
+    @Test
+    @DisplayName("redo stack must be empty after redo operation")
+    public void redoStackEmptyAfterRedoOperation() {
+        StringBuilder text = new StringBuilder("Text");
+        CommandManager<StringBuilder> commander = new CommandManager<>(text);
+        commander.execute(new TestCommand());
+        commander.undo();
+        commander.redo();
+        Assertions.assertTrue(commander.isRedoStackEmpty());
+    }
+
+    @Test
+    @DisplayName("undo stack must be not empty after redo operation")
+    public void undoStackNotEmptyAfterRedoOperation() {
+        StringBuilder text = new StringBuilder("Text");
+        CommandManager<StringBuilder> commander = new CommandManager<>(text);
+        commander.execute(new TestCommand());
+        commander.undo();
+        commander.redo();
+        Assertions.assertFalse(commander.isUndoStackEmpty());
+    }
+
+    @Test
+    @DisplayName("runnable called when command executed")
+    public void runnableIsCalledWhenCommandIsExecuted() {
+        StringBuilder text = new StringBuilder("Text");
+        AtomicBoolean aBoolean = new AtomicBoolean();
+        CommandManager<StringBuilder> commander = new CommandManager<>(text, () -> aBoolean.set(!aBoolean.get()));
+        commander.execute(new TestCommand());
+        Assertions.assertTrue(aBoolean.get());
+    }
+
+    @Test
+    @DisplayName("runnable called when undo is executed")
+    public void runnableIsCalledWhenUndoIsExecuted() {
+        StringBuilder text = new StringBuilder("Text");
+        AtomicBoolean aBoolean = new AtomicBoolean();
+        CommandManager<StringBuilder> commander = new CommandManager<>(text, () -> aBoolean.set(!aBoolean.get()));
+        commander.execute(new TestCommand());
+        commander.undo();
+        Assertions.assertFalse(aBoolean.get());
+    }
+
+    @Test
+    @DisplayName("runnable called when redo is executed")
+    public void runnableIsCalledWhenRedoIsExecuted() {
+        StringBuilder text = new StringBuilder("Text");
+        AtomicBoolean aBoolean = new AtomicBoolean();
```

Comment:
The `AtomicBoolean` can be a `AtomicInteger` and we could increment it in every call to the runnable, therefore being able to test the value as 1, 2 and 3 in `runnableIsCalledWhenCommandIsExecuted`, `runnableIsCalledWhenUndoIsExecuted` and `runnableIsCalledWhenRedoIsExecuted` respectively. I leave it for you to decide.

Reply:
Makes sense, you would be testing that runnable is called on every command operation

Reply:
Pushed